// @garmin/fitsdk/src/bit-stream.js
/////////////////////////////////////////////////////////////////////////////////////////////
// Copyright 2026 Garmin International, Inc.
// Licensed under the Flexible and Interoperable Data Transfer (FIT) Protocol License; you
// may not use this file except in compliance with the Flexible and Interoperable Data
// Transfer (FIT) Protocol License.
/////////////////////////////////////////////////////////////////////////////////////////////
// ****WARNING****  This file is auto-generated!  Do NOT edit this file.
// Profile Version = 21.205.0Release
// Tag = production/release/21.205.0-0-gb3c261eb
/////////////////////////////////////////////////////////////////////////////////////////////


import FIT from "./fit.js";

class BitStream {
    #data = null;
    #array = null;
    #currentArrayPosition = 0;
    #bitPerPosition = 0;
    #isBigInt = false;
    #currentByte = 0;
    #currentBit = 0;
    #bitsAvailable = 0;

    constructor(data, baseType = FIT.BaseType.UINT8) {
        this.#data = Array.isArray(data) ? data : [data];
        const baseTypeSize = FIT.BaseTypeDefinitions[baseType].size;
        this.#bitPerPosition = baseTypeSize * 8;
        this.#isBigInt = this.#bitPerPosition > 32;
        this.reset();
    }

    get bitsAvailable() {
        return this.#bitsAvailable;
    }

    get hasBitsAvailable() {
        return this.#bitsAvailable > 0;
    }

    reset() {
        this.#array = null;
        this.#currentArrayPosition = 0;
        this.#bitsAvailable = this.#bitPerPosition * this.#data.length;
        this.#currentByte = 0;
        this.#currentBit = 0;
    }

    readBit() {
        if (!this.hasBitsAvailable) {
            this.#throwError();
        }

        if (this.#array === null) {
            this.#array = this.#data;
            this.#currentArrayPosition = this.#data.length - (this.#bitsAvailable / this.#bitPerPosition);
            this.#nextByte();
        }

        if (this.#currentBit >= this.#bitPerPosition) {
            this.#nextByte();
        }

        let bit;
        if (this.#isBigInt) {
            bit = Number(this.#currentByte & 1n);
            this.#currentByte >>= 1n;
        } else {
            bit = this.#currentByte & 0x01;
            this.#currentByte >>= 1;
        }
        this.#currentBit++;
        this.#bitsAvailable--;

        return bit;
    }

    readBits(nBitsToRead) {
        // Fast path: reading exactly one value's worth of bits and #array not yet initialized
        if (nBitsToRead === this.#bitPerPosition && this.#array === null) {
            if (!this.hasBitsAvailable) {
                this.#throwError();
            }
            const index = this.#data.length - (this.#bitsAvailable / this.#bitPerPosition);
            const value = this.#data[index];
            this.#bitsAvailable -= nBitsToRead;

            if (this.#isBigInt) {
                return Number(BigInt.asUintN(nBitsToRead, value));
            }
            if (nBitsToRead < 32) {
                return (value & ((1 << nBitsToRead) - 1)) >>> 0;
            }
            return value >>> 0;
        }

        if (nBitsToRead <= 32) {
            let value = 0;
            for (let i = 0; i < nBitsToRead; i++) {
                value |= this.readBit() << i;
            }
            return value >>> 0;
        }

        let value = 0n;

        for (let i = 0, shift = 0n; i < nBitsToRead; i++, shift++) {
            value |= BigInt(this.readBit()) << shift;
        }

        return Number(value);
    }

    #nextByte() {
        if (this.#currentArrayPosition >= this.#array.length) {
            this.#throwError();
        }

        this.#currentByte = this.#array[this.#currentArrayPosition++];
        this.#currentBit = 0;
    }

    get array() {
        return this.#array;
    }

    #throwError(error = "") {
        throw Error("FIT Runtime Error no bits available.");
    }
}

export default BitStream;
