// @garmin/fitsdk/src/stream.js
/////////////////////////////////////////////////////////////////////////////////////////////
// Copyright 2026 Garmin International, Inc.
// Licensed under the Flexible and Interoperable Data Transfer (FIT) Protocol License; you
// may not use this file except in compliance with the Flexible and Interoperable Data
// Transfer (FIT) Protocol License.
/////////////////////////////////////////////////////////////////////////////////////////////
// ****WARNING****  This file is auto-generated!  Do NOT edit this file.
// Profile Version = 21.205.0Release
// Tag = production/release/21.205.0-0-gb3c261eb
/////////////////////////////////////////////////////////////////////////////////////////////


import FIT from "./fit.js";
import UtilsInternal from "./utils-internal.js";

class Stream {
    #position = 0;
    #arrayBuffer = null;
    #dataView = null;
    #textDecoder = new TextDecoder("utf-8", { fatal: false, ignoreBOM: true });
    #crcCalculator = null;
    #legacyArrayMode = false;

    static fromByteArray(data) {
        const buf = new Uint8Array(data);
        return this.fromArrayBuffer(buf.buffer);
    }

    static fromBuffer(buffer) {
        const arrayBuffer = buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);
        return this.fromArrayBuffer(arrayBuffer);
    }

    static fromArrayBuffer(arrayBuffer) {
        const stream = new Stream(arrayBuffer);
        return stream;
    }

    constructor(arrayBuffer) {
        this.#position = 0;
        this.#arrayBuffer = arrayBuffer;
        this.#dataView = new DataView(arrayBuffer);
    }

    get length() {
        return this.#arrayBuffer.byteLength;
    }

    get bytesRead() {
        return this.#position;
    }

    get position() {
        return this.#position;
    }

    get crcCalculator() {
        return this.#crcCalculator;
    }

    set crcCalculator(crcCalculator) {
        this.#crcCalculator = crcCalculator;
    }

    get legacyArrayMode() {
        return this.#legacyArrayMode;
    }

    set legacyArrayMode(legacyArrayMode) {
        this.#legacyArrayMode = legacyArrayMode;
    }

    reset() {
        this.seek(0);
    }

    seek(position) {
        this.#position = position;
    }

    slice(begin, end) {
        return this.#arrayBuffer.slice(begin, end);
    }

    peekByte() {
        return this.#dataView.getUint8(this.#position);
    }

    readByte() {
        return this.readUInt8();
    }

    readBytes(size) {
        if (this.#position + size > this.#arrayBuffer.byteLength) {
            throw Error(`FIT Runtime Error end of stream at byte ${this.#position}`);
        }

        const bytes = new Uint8Array(this.#arrayBuffer, this.#position, size);
        this.#position += size;

        this.#crcCalculator?.addBytes(bytes, 0, size);

        return bytes;
    }

    readUInt8(opts) {
        return this.readValue(FIT.BaseType.UINT8, 1, { convertInvalidToNull: false, ...opts });
    }

    readInt8(opts) {
        return this.readValue(FIT.BaseType.SINT8, 1, { convertInvalidToNull: false, ...opts });
    }

    readUInt16(opts) {
        return this.readValue(FIT.BaseType.UINT16, 2, { convertInvalidToNull: false, ...opts });
    }

    readInt16(opts) {
        return this.readValue(FIT.BaseType.SINT16, 2, { convertInvalidToNull: false, ...opts });
    }

    readUInt32(opts) {
        return this.readValue(FIT.BaseType.UINT32, 4, { convertInvalidToNull: false, ...opts });
    }

    readInt32(opts) {
        return this.readValue(FIT.BaseType.SINT32, 4, { convertInvalidToNull: false, ...opts });
    }

    readUInt64(opts) {
        return this.readValue(FIT.BaseType.UINT64, 8, { convertInvalidToNull: false, ...opts });
    }

    readInt64(opts) {
        return this.readValue(FIT.BaseType.SINT64, 8, { convertInvalidToNull: false, ...opts });
    }

    readFloat32(opts) {
        return this.readValue(FIT.BaseType.FLOAT32, 4, { convertInvalidToNull: false, ...opts });
    }

    readFloat64(opts) {
        return this.readValue(FIT.BaseType.FLOAT64, 8, { convertInvalidToNull: false, ...opts });
    }

    readString(strlen, opts = {}) {
        return this.readValue(FIT.BaseType.STRING, strlen, opts);
    }

    asUint8Array() {
        return new Uint8Array(this.#arrayBuffer);
    }

    readValue(
        baseType,
        size,
        {
            littleEndian = true,
            convertInvalidToNull = true,
            isArray = false
        } = {}) {
        const baseTypeDef = FIT.BaseTypeDefinitions[baseType];
        const baseTypeSize = baseTypeDef.size;
        const baseTypeInvalid = baseTypeDef.invalid;

        const startPos = this.#position;
        const bytes = this.readBytes(size);

        if (size % baseTypeSize !== 0) {
            return convertInvalidToNull ? null : baseTypeInvalid;
        }

        if (baseType === FIT.BaseType.STRING) {
            const string = this.#textDecoder.decode(bytes).replace(/\uFFFD/g, "");
            const strings = string.split(FIT.NULL_TERMINATOR);

            while (strings[strings.length - 1] === "") {
                strings.pop();
            }

            if (strings.length === 0) {
                return null;
            }

            if (this.#legacyArrayMode) {
                return strings.length === 1 ? strings[0] : strings;
            }

            return strings[0];
        }

        const count = size / baseTypeSize;

        if (count === 1) {
            const value = this.#readSingleValue(baseType, startPos, littleEndian);

            if (convertInvalidToNull && value === baseTypeInvalid) {
                return null;
            }

            if (!this.#legacyArrayMode && !isArray) {
                return value;
            }

            return UtilsInternal.sanitizeValues([value], { legacyArrayMode: this.#legacyArrayMode, isArray });
        }

        let values = [];

        for (let i = 0; i < count; i++) {
            values.push(this.#readSingleValue(baseType, startPos + i * baseTypeSize, littleEndian));
        }

        if (convertInvalidToNull) {
            if (UtilsInternal.onlyInvalidValues(values, baseTypeInvalid)) {
                return null;
            }

            if (baseType !== FIT.BaseType.BYTE) {
                values = values.map(value => value === baseTypeInvalid ? null : value);
            }
        }

        return UtilsInternal.sanitizeValues(
            values,
            {
                legacyArrayMode: this.#legacyArrayMode,
                isArray
            });
    }

    #readSingleValue(baseType, offset, littleEndian) {
        switch (baseType) {
            case FIT.BaseType.BYTE:
            case FIT.BaseType.ENUM:
            case FIT.BaseType.UINT8:
            case FIT.BaseType.UINT8Z:
                return this.#dataView.getUint8(offset);

            case FIT.BaseType.SINT8:
                return this.#dataView.getInt8(offset);

            case FIT.BaseType.UINT16:
            case FIT.BaseType.UINT16Z:
                return this.#dataView.getUint16(offset, littleEndian);

            case FIT.BaseType.SINT16:
                return this.#dataView.getInt16(offset, littleEndian);

            case FIT.BaseType.UINT32:
            case FIT.BaseType.UINT32Z:
                return this.#dataView.getUint32(offset, littleEndian);

            case FIT.BaseType.SINT32:
                return this.#dataView.getInt32(offset, littleEndian);

            case FIT.BaseType.UINT64:
            case FIT.BaseType.UINT64Z:
                return this.#dataView.getBigUint64(offset, littleEndian);

            case FIT.BaseType.SINT64:
                return this.#dataView.getBigInt64(offset, littleEndian);

            case FIT.BaseType.FLOAT32:
                return this.#dataView.getFloat32(offset, littleEndian);

            case FIT.BaseType.FLOAT64:
                return this.#dataView.getFloat64(offset, littleEndian);

            default:
                throw new Error(`Unsupported FIT base type: ${baseType}`);
        }
    }
}

export default Stream;
